### DISCUSSION TITLE
Merging dsgibbons/shap

### DISCUSSION THREAD

**Discussion by dsgibbons** [shap/shap#2942] [General] [answered]
I've been working with a couple of collaborators (most notably @thatlittleboy and @connortann) on a fork of SHAP at https://github.com/dsgibbons/shap. We have made several quality-of-life improvements around setup, testing, and warnings. @slundberg has made the three of us collaborators here at slundberg/shap. 

The purpose of this thread is to discuss how we should best start merging changes from dsgibbons/shap. Do we want to raise individual PRs for each fix? Or just one PR (without squashing so that our incremental changes can be seen). It appears that merging is blocked until all tests pass, so perhaps a single PR is a better way to go. As we go through this process, we should close relevant issues and PRs to start clearing out the backlog.

**Comment by connortann** [comment]
Fantastic news - glad that we can come together in a single repo.

Within reason, I think multiple PRs would be a preferable way to go as it allows a clearer audit trail of each decision, and allows the PRs to be linked back to the fork PRs for discussion and comments.

One way to start could be creating a **minimum set of changes required to get the tests passing**, and using that as the basis for the first major bugfix PR. I could have a go at that on a branch in the fork.

Then, once the tests are passing we can raise PRs for other changes such as:

- Linting config & fixes
- Fixing warnings
- CI configuration
- Adding python 3.11
- Changes to docs

How does that sound?

**Reply by dsgibbons** [reply]
> One way to start could be creating a minimum set of changes required to get the tests passing, and using that as the basis for the first major bugfix PR. I could have a go at that on a branch in the fork.

I think that's a good idea.

@slundberg has also mentioned the possibility of standing up a shap organisation (i.e., move slundberg/shap to shap/shap). I mentioned this to @thatlittleboy, and he thinks that we should set up the organisation first before merging. What do you think @connortann? I'm not sure how long it will take to get an organisation set up. If it's going to be a while, I wouldn't mind starting to merge sooner rather than later.

**Reply by thatlittleboy** [reply]
I've not done this before, but presumably we can go with this workflow: https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository#transferring-a-repository-owned-by-your-personal-account.

@dsgibbons Did @slundberg mention who the owner of the organization would be, yourself or himself? merging the changes from the dsgibbons/shap fork to here is a semi-large task, which is why I'm advocating for migrating the repo first. (I'm expecting the former to take more effort than the latter).

In the meantime, we still have some open PRs over at dsgibbons/shap (I won't open anymore :p), we should also clear those out before doing the complete migration from dsgibbons -> here / shap org.

Lastly, to @connortann 's suggestion, I'm thinking that cherry picking the minimal fixes, rather than following the natural time/commit progression of `dsgibbons/shap` repo, would likely result in plenty of merge conflicts. And if we *then* have to fix those conflicts manually, it'll just increase the likelihood of us making mistakes.  
I think putting everything into one PR (and not squashing) would likely be a cleaner approach.

**Reply by connortann** [reply]
After some experimentation over the past 30 mins, I agree your original suggestion of making a single PR is probably the best route to go down. I looked into cherry picking a subset of changes: this approach does indeed generate a large number of conflicts. So, we could create a new branch on top of master to revert any unwanted changes, then make a single PR.

**Reply by connortann** [reply]
To the point about creating an organisation: it sounds like a good idea in principle. However, I'm rather cautious about causing disruption, for example to any external resources that point to the current URL. What problems are we hoping to address by migrating to an org?

**Reply by connortann** [reply]
Regarding merges, I tentatively made a draft PR between forks here, to explore if the CI will pass: https://github.com/slundberg/shap/pull/2943

At least one of the checks fails (the CodeQL C++ check), so without changing the repository settings we won't be able to merge. EDIT: fixed now.

Once the CI passes though, I think we'd actually be quite close to merging the forks. We would want to edit the description to link all open Issues that would be closed by the PR. What else would be required other than that?

**Reply by thatlittleboy** [reply]
> What problems are we hoping to address by migrating to an org?

I think @connortann brings up a good point, and we should deliberate this carefully. I'll start with some background:

* Repos on **personal accounts** (i.e., `slundberg/shap`) only have 2 permission levels (Owner, and Collaborators). And there can only be _one owner_. [Source](https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-personal-account-settings/permission-levels-for-a-personal-account-repository#about-permissions-levels-for-a-personal-account-repository)
* A collaborator has write/push access to a repository, but not much else. Compare the Owner v.s. Collaborator access [here](https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-personal-account-settings/permission-levels-for-a-personal-account-repository#owner-access-for-a-repository-owned-by-a-personal-account)
* And we faced this first-hand in `dsgibbons/shap` -- e.g. I can't edit the repository to add the documentation link to the repo sidebar, only @dsgibbons could. Things like codecov that requires managing keys, etc. have to be done by the owner, not the collaborators.
* Members of an **organization** (let's say, `shap/shap`), on the other hand, can have a variety of and more granular access control. It is in general, less restrictive than Collaborators. [Source](https://docs.github.com/en/organizations/collaborating-with-groups-in-organizations/about-organizations#about-organizations), in particular read the [Roles in an organization](https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/repository-roles-for-an-organization).

With that background (^), 

* I think having an organization is more in-line with our original goal, where we don't have a singular person being the chokepoint of the progress of a project. Sure, as collaborators we can still merge PRs and make releases, but what if we want to invite more collaborators? Or manage repository secrets? Supposing one day the owner of the Personal Account repo goes MIA, we'll be back at square one again (forced to create forks). So, for the long-term future of an established project, I do believe setting up an organization is perhaps a wiser choice.
* W.r.t. links being broken, etc.: here's a link describing what happens when we **transfer** a Personal account repo to an organization repo: https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository#whats-transferred-with-a-repository. 
  ![image](https://github.com/slundberg/shap/assets/30731072/1e128e31-5a05-4af5-bae8-221022170e46)

**Reply by connortann** [reply]
Thanks for the detailed response. The more granular permissions seem like very important tools in cultivating a larger pool of maintainers, so it seems worth it in the long run.

In parallel, I think we can also get the fork ready for merging.

**Reply by connortann** [reply]
I created a meta-issue for tracking progress towards merging the forks: https://github.com/slundberg/shap/issues/2944 .  Please feel free to edit the TODO and add anything I've missed!

**Reply by connortann** [reply]
I'm glad to say PR #2943 is passing, so for myself I feel we're ready to merge 🎉! However, I've kept the PR as a "Draft" until we have a concensus here that we should proceed. @dsgibbons & @thatlittleboy , what do you think?

We haven't migrated to a GitHub Org yet which @thatlittleboy advocated doing first - so, just wanted to check if this is something you think we should wait for?

**Reply by dsgibbons** [reply]
> @dsgibbons Did @slundberg mention who the owner of the organization would be, yourself or himself? merging the changes from the dsgibbons/shap fork to here is a semi-large task, which is why I'm advocating for migrating the repo first. (I'm expecting the former to take more effort than the latter).

No, not yet. My first assumption was that @slundberg would start/own the organisation. I will seek some clarity on this if he does not respond to this discussion.

> In the meantime, we still have some open PRs over at dsgibbons/shap (I won't open anymore :p), we should also clear those out before doing the complete migration from dsgibbons -> here / shap org.

Agreed.

> I think having an organization is more in-line with our original goal, where we don't have a singular person being the chokepoint of the progress of a project. Sure, as collaborators we can still merge PRs and make releases, but what if we want to invite more collaborators? Or manage repository secrets? Supposing one day the owner of the Personal Account repo goes MIA, we'll be back at square one again (forced to create forks). So, for the long-term future of an established project, I do believe setting up an organization is perhaps a wiser choice.

Great response re: the benefits of having an organisation. It would be great to avoid the singular-person problem in the future. We would want a non-free plan to get access to standard features like protected branches and draft pull requests. It remains to be seen how this would be funded. I'm hoping for more details from @slundberg soon.

**Reply by thatlittleboy** [reply]
> We would want a non-free plan to get access to standard features like protected branches and draft pull requests

I *think* those restrictions are only for private repositories. As long as we keep shap public , it should be fine? Source: https://docs.github.com/en/get-started/learning-about-github/githubs-products#github-free-for-organizations

**Reply by dsgibbons** [reply]
Good spot. The pricing page didn't make that so clear. Well that should make it easier to get the organisation off the ground!

**Reply by connortann** [reply]
@dsgibbons @slundberg great news. Do you have an expected timeline for setting up the shap org?

**Reply by dsgibbons** [reply]
We have one already 😉 you should receive an invite soon @connortann!

**Reply by connortann** [reply]
Thanks, joined. It would be great to discuss the process for making a release: for example, who can trigger one, how relevant Secrets will be managed, what checks need to be made, and so on. We could start by creating a new discussion or meta-issue.

**Reply by connortann** [reply]
@dsgibbons @thatlittleboy shall we create a live chat  somewhere for shap maintainers, for example on Discord / Slack / Gitter? Or, would you prefer just to chat on GitHub discussions & issues?

**Reply by dsgibbons** [reply]
I'm happy with GitHub discussions - unless we need a private channel for managing secrets?

**Reply by connortann** [reply]
OK, fine with me. A benefit is that the discussion is public and we can encourage more people to join!

**Comment by dsgibbons** [comment]
> I'm glad to say PR https://github.com/slundberg/shap/pull/2943 is passing, so for myself I feel we're ready to merge tada! However, I've kept the PR as a "Draft" until we have a concensus here that we should proceed. @dsgibbons & @thatlittleboy , what do you think?

That's awesome! I'm happy to merge now. I think the organisation may take a while to get going.

**Reply by thatlittleboy** [reply]
Agreed, let's do the merge first then. (Reviewing now)

**Reply by connortann** [reply]
Any update on the migration to an Org @dsgibbons @slundberg ?

Here's a concise list of actions that we need admin permissions to address:

[ Edited ]

1. Releases
   - [X] Add TestPyPi token for #3014
   - [x] Give us the thumbs up to make a release once items in #2979 are closed
2. CI settings
   - [x] Consider disabling the "required Checks" requirement, so pytest can be skipped on PRs that do not change the package code
   - [x] Consider disabling the _"Require branches to be up to date before merging"_ requirement, as it leads to a lot of often unneccessary extra builds (each PR has to be rebased and build before merging, so it takes over an hour to complete a handful of PRs)
   - [x] Re-activate ReadtheDocs build on PRs so we can better review doc-related PRs. #3038
   - [x] Add the documentation page link to the GH front page sidebar,
3. Misc
   - [x] Add the `good first issue` label, as recommended by [github](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/encouraging-helpful-contributions-to-your-project-with-labels)
4. Handover
   - [x] Migrate to a SHAP org
   - [x] Add as Owners in PyPi and TestPyPi
   - [x] Add as owners in ReadTheDocs.com

**Reply by thatlittleboy** [reply]
I'll add on to @connortann 's list,

4. Re-activate ReadtheDocs build on PRs, as well as adding the documentation page link to the GH front page sidebar, so we can better review doc-related PRs. (#3038)

**Reply by connortann** [reply, marked as answer]
@dsgibbons  @thatlittleboy  just FYI, everything on the the todo list above has now been implemented 🎉

**Comment by connortann** [comment]
Once #2943 is merged (and it looks like we're pretty much there), I think we need a way to co-ordinate on our priorities as maintainers. Shall we sett up a GH [Project](https://github.com/slundberg/shap/projects?type=classic) (classic) to help us triage things?

Some priorities off the top of my head:

- [x] Get pre-commit running on CI again
- [x] Dependabot update: #2843

[Edited for brevity]

**Reply by thatlittleboy** [reply]
I'm open to giving Projects a try. Would we just use the same Projects that slundberg has created from before?  
We can also mark issues/PRs with labels/milestones + having an assignee to signify that we're working on a particular issue, as a quick and dirty solution for now.

For myself, my immediate priority is still to attempt to resolve some of the bugs that have been in the issue tracker for a long time first. Knock out as many of these as we can before the next imminent release.

**Reply by connortann** [reply]
I created a milestone: [2023-06-fixes](https://github.com/slundberg/shap/milestone/1), which we can use to mark some of the issues and PRs that we think should be addressed in the next few weeks.

**Comment by jklaise** [comment]
As a developer of [alibi](https://github.com/SeldonIO/alibi) I'd like to extend my thanks to all of you for pushing forward with the community fork and getting the much needed fixes merged upstream, here's waiting for the next release!

One question from me, now that you have been made collaborators here, what would be the future of the community fork? Should external contributors now report issues and open PRs here (upstream)? Perhaps a small notice on the community fork would be helpful. I also realize that the current state of the issue tracker here is unwieldy and may require some strategizing over how to make it manageable again.

**Reply by connortann** [reply]
Yes, I think external contributors should now open issues and PRs here. Yep, let's update the fork with a notice that it has been successfully merged in.

FYI, We are investigating creating a GitHub Organization for SHAP to allow a larger pool of maintainers, so we might have slower progress on managing PRs until that is done due to various permissions limitations.

You're definitely right about managing the issue tracker. We need to set up a better way of prioriting issues and communicating between maintainers: for example, using milestones, projects, and potentially using some automation to mark old issues as stale.
